### DISCUSSION TITLE
Sometimes failing intune Rollout

### DISCUSSION THREAD

**Discussion by sschmocker** [Romanitho/Winget-Install#31] [General]
We are using your script to push softwares via intune. we Also push the appinstaller itself via appx file in intune to make sure every device got winget. 
Here is an example of whats happening on a freshly wiped device.
Any Idea whats happening there?

Here it cant find the application:
![2022-04-25 13_57_22-Software deployment monitoring4](https://user-images.githubusercontent.com/96051818/165084941-fe4c6941-7f09-4266-83fd-7f780df8bcd7.png)

10 minutes later it works fine:
![2022-04-25 13_57_22-Software deployment monitoring3](https://user-images.githubusercontent.com/96051818/165084947-cfa31e43-37f0-41b1-b7ef-38c94c118c7e.png)

finds the application but fails:
![2022-04-25 13_57_22-Software deployment monitoring](https://user-images.githubusercontent.com/96051818/165084710-4e4655cf-b2c5-4859-b344-8676c9030cc4.png)

Works fine after a retry couple minutes later:
![2022-04-25 13_57_22-Software deployment monitoring2](https://user-images.githubusercontent.com/96051818/165084707-dd08a736-5ade-4e5c-bd76-2aaa2e472607.png)

**Comment by Romanitho** [comment]
Hi. We have identified an issue on the current Winget version (1.2)
You can install version 1.3 (The latest preview version on the MS github) https://github.com/microsoft/winget-cli/releases/tag/v1.3.431

Try this and let us know ;)

**Reply by sschmocker** [reply]
thx for your reply.
Same Problem with latest preview:
![2022-04-25 13_57_22-Software deployment monitoring5](https://user-images.githubusercontent.com/96051818/165092333-e09cd7e4-3c77-44a6-b8e1-9b3c15a27718.png)

**Reply by Romanitho** [reply]
The thing is, if appinstaller  is pushed with intune, not sure it is applied straight away to all users. this can explain why you have a delay.

**Reply by Romanitho** [reply]
what you can try is to first install WAU (as it has all the prerequisites)

**Reply by sschmocker** [reply]
Well the appinstaller was already on the device, i quickly checked that via powershell once the device was on the desktop. 
WAU is not an option for me, because we are not using the updater. We just use the winget-install script for appdeployments. Maybe a prerequisites installation would make sense in the winget-installer script

**Reply by Romanitho** [reply]
With Intune, I use dependency package instead. That allow us to keep clear Install scripts.

In the AllInOne Repo, I use this function to check prerequisites : 

```
function Get-WingetStatus{
    Write-Host -ForegroundColor yellow "Checking prerequisites..."
    $hasAppInstaller = Get-AppXPackage -Name 'Microsoft.DesktopAppInstaller'
    [Version]$AppInstallerVers = $hasAppInstaller.version
    if ($AppInstallerVers -gt "1.18.0.0"){
        Write-Host -ForegroundColor Green "WinGet is already installed."
    }
    else {
        Write-Host -ForegroundColor Red "WinGet missing."
        Write-Host -ForegroundColor Yellow "-> Installing WinGet prerequisites..."

        #installing dependencies
        $ProgressPreference = 'SilentlyContinue'
        if (Get-AppxPackage -Name 'Microsoft.UI.Xaml.2.7'){
            Write-Host -ForegroundColor Green "-> Prerequisite: Microsoft.UI.Xaml.2.7 exists"
        }
        else{
            Write-Host -ForegroundColor Yellow "-> Prerequisite: Installing Microsoft.UI.Xaml.2.7"
            $UiXamlUrl = "https://www.nuget.org/api/v2/package/Microsoft.UI.Xaml/2.7.0"
            Invoke-RestMethod -Uri $UiXamlUrl -OutFile ".\Microsoft.UI.XAML.2.7.zip"
            Expand-Archive -Path ".\Microsoft.UI.XAML.2.7.zip" -DestinationPath ".\extracted" -Force
            Add-AppxPackage -Path ".\extracted\tools\AppX\x64\Release\Microsoft.UI.Xaml.2.7.appx"
            Remove-Item -Path ".\Microsoft.UI.XAML.2.7.zip" -Force
            Remove-Item -Path ".\extracted" -Force -Recurse
        }

        Write-Host -ForegroundColor Yellow "-> Prerequisite: Installing Microsoft.VCLibs.x64.14.00.Desktop"
        Add-AppxPackage -Path https://aka.ms/Microsoft.VCLibs.x64.14.00.Desktop.appx

        #installin Winget
        Write-Host -ForegroundColor Yellow "-> Installing Winget..."
        Add-AppxPackage -Path https://github.com/microsoft/winget-cli/releases/download/v1.3.431/Microsoft.DesktopAppInstaller_8wekyb3d8bbwe.msixbundle


        $hasAppInstaller = Get-AppXPackage -Name 'Microsoft.DesktopAppInstaller'
        [Version]$AppInstallerVers = $hasAppInstaller.version
        if ($AppInstallerVers -gt "1.18.0.0"){
            Write-Host -ForegroundColor Green "WinGet successfully installed."
        }
        else{
            Write-Host -ForegroundColor Red "WinGet failed to installed."
        }
    }
}
```

Only Visual c++ 2022 is installed beside.

**Reply by sschmocker** [reply]
I actually tried that right now, for every App i used Visual C++ as dependency. 
First try, all works as expected. (Like usual with your script :D )

**Reply by sschmocker** [reply]
Well it seems that it is not 100% fixed:

![image](https://user-images.githubusercontent.com/96051818/165704475-7cc64cd9-50f4-4eee-abc0-327d143908d7.png)

**Reply by Romanitho** [reply]
is it only the case for Adobe ? or all your apps ?

**Reply by sschmocker** [reply]
yes its just for Adobe. Sometimes it works sometimes not.
I also created a requirement script to make sure that the minimum version of appinstaller needs to be on the device. Here the script:

$version = Get-ItemProperty -Path "HKCU:\Software\Classes\Local Settings\Software\Microsoft\Windows\CurrentVersion\AppModel\Repository\Packages\Microsoft.Desktop*" | Select-Object -ExpandProperty PackageID
if ($version -gt "Microsoft.DesktopAppInstaller_1.18.000.0_x64__8wekyb3d8bbwe"){
Write-Output "0"
}

![image](https://user-images.githubusercontent.com/96051818/165782207-4d53599a-4644-4ced-b7ee-a5e6c477f4db.png)

**Reply by Romanitho** [reply]
Visual C++ 2019 or 2022 needs to be installed as well.
I need to make logs better on the Winget-Install script. This is something I need to work on.

**Reply by sschmocker** [reply]
yes i know, I made Visual C++ as a dependency package for every app im deploying. 
Better logging would be great, we use your script in combination with a teams webhook so we have logging in teams channels. 
If anyone is interested we use it in combination with that:
https://scloud.work/en/powershell-win32-log-to-teams-intune/?fbclid=IwAR2tu4kA7kIG9ZvtqTauRl47-tpO5Ta5GeFO0Yg5yKQC80UoGaCzrq7eJ1Q

**Reply by Romanitho** [reply]
I will probably not have time to work on that this weekend.
And thanks for sharing, it's something i'll definitely test

**Reply by sschmocker** [reply]
Adobe still shows failed for us. But it seems that it gets installed tho. 
I now choose to supress toast notifications so users dont get confused.
